### DISCUSSION TITLE
Cant export modified model-viewer

### DISCUSSION THREAD

**Discussion by discodevl** [google/model-viewer#3048] [Q&A]
Im making a code to export an modified glb model, but it only returns me the original model with no modifications.
```
let options = {
      binary: true,
      trs: true,
      onlyVisible: true,
      maxTextureSize: 256,
      forcePowerOfTwoTextures: false,
      includeCustomExtensions: false,
      embedImages: true
    };

const newGlb = await document.querySelector("model-viewer").exportScene(options);

let file = new File([newGlb], "newVariant.glb");
let link = document.createElement("a");
link.download = file.name;
link.href = URL.createObjectURL(file);
link.click();

```
This code runs after I change a texture of the modelViewer, it downloads the glb with no modifications. I expect to download the glb with the texture I setted before.

**Comment by elalish** [comment]
Can  you provide a link to a repro URL? What you describe should work (our editor does exactly this). How are you changing textures?

**Reply by discodevl** [reply]
```
async function changeTexture(event) {
    const materialsViewer =
      document.querySelector("model-viewer").model.materials;

    const selectedMaterialViewer = materialsViewer.find(
      (mat) => mat.name === props.materialSelected.name
    );

    try {
      selectedMaterialViewer.pbrMetallicRoughness.baseColorTexture.texture.source.setURI(URL.createObjectURL(event.target.files[0]));
    } catch {
      let texture = document.querySelector("model-viewer").createTexture(URL.createObjectURL(event.target.files[0]));
      selectedMaterialViewer.pbrMetallicRoughness.baseColorTexture.setTexture(texture);
    }

    let options = {
      binary: true,
      trs: true,
      onlyVisible: true,
      maxTextureSize: Infinity,
      forcePowerOfTwoTextures: false,
      includeCustomExtensions: false,
      embedImages: true,
    };

    const newGlb = await document.querySelector("model-viewer").exportScene(options);
    let file = new File([newGlb], "newVariant.glb");
    let link = document.createElement("a");
    link.download = file.name;
    link.href = URL.createObjectURL(file);
    link.click();
}

//JSX CODE:
<div >
    <input type="file" accept="image/*" onChange={changeTexture} />
</div>
```
result:
1- this is the loaded glb
![firstGlbLoaded](https://user-images.githubusercontent.com/69987663/145860197-c9bc23f4-6c5e-4927-9a11-f93d30c9a34e.png)

2- Apply the new texture and download the glb "newVariant.glb" (supposed to be this black astronaut)
![modfiedGlb](https://user-images.githubusercontent.com/69987663/145860403-93b9ef8a-200e-4123-865b-dddfe22211d0.png)

3- This is the newVariant.glb 
![Screenshot from 2021-12-13 18-29-34](https://user-images.githubusercontent.com/69987663/145861020-6bf7df8a-9ea2-489c-ac15-1c38bc31416c.png)
![newVariantGlb](https://user-images.githubusercontent.com/69987663/145860721-b361a191-acd8-48d1-8899-56039c1fac78.png)

**Reply by discodevl** [reply]
Update:
It works only if change other texture except the baseColorTexture

**Reply by slightHYP** [reply]
Do you enter in the `try` or in the `catch` block? 
If you enter inside the `catch`, the `createTexture` function returns a Promise and I think that you have to wait for its return before set the texture.

Like in this example:
https://modelviewer.dev/examples/scenegraph/#createTexturesExample

Without `await` it is possible that the `exportScene` is executed before the new texture is applied, but this is just an hypothesis without debugging

**Comment by discodevl** [comment]
Problem solved!
I moved the export code outside the changeTexture function now I can download the glb with the modifications.
Thanks for the answers!
